Feature: STM32 DESig UID decoder

ALTracer opened the pull request:
## Detailed description

* This is a new feature expanding the existing stm32{h7,mp15} (potentially h5) UID-printing targe commands.
* The existing problems are plain hex in multiple O-packet and unused pattern/decode info.
* This PR solves these by buffering hex into a stack `char[]` (sprintf) and adding a small parser/decoder based on info from RMs.

TODO includes migrating sprintf usage to utoa (upper or lower) and testing the decoder on other families like STM32F4/F2/F7 which also include a Digital Electronic Signature containing Lot and Wafer info.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (see [Building the firmware](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-firmware))
* [x] It builds as BMDA (see [Building the BMDA](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-app))
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

## Closing issues

dragonmux commented on src/target/stm32h7.c:
```diff
@@ -457,6 +457,18 @@ static bool stm32h7_mass_erase(target_s *target)
+ * Unique device ID register
+ * Other ref.manuals omit the decoding details.
+ */
+struct __attribute__((packed)) stm32h7_uid {
```
This needs the `typedef`'ing treatment like the one in stm32mp15.c

dragonmux commented on src/target/stm32h7.c:
```diff
@@ -473,15 +485,22 @@ static bool stm32h7_uid(target_s *target, int argc, const char **argv)
 	(void)argc;
 	(void)argv;
 
+	struct stm32h7_uid uid = {0};
```
This likewise wants moving down to the memcpy and the `= {0}` optionally dropped same as the stm32mp15.c code

dragonmux reviewed (changes requested):
This is definitely getting there.

To fix the windows builds, the packing attribute needs defining in general.h - `__attribute__((packed))` for GCC and Clang, and you'll have to deal with the [packing pragmas](https://learn.microsoft.com/en-us/cpp/preprocessor/pack?view=msvc-170) for MSVC.

ALTracer commented:
Revisiting. I dropped the duplicate changes in stm32h7.c and developed functionality in the stm32mp15.c driver just so I can be DRY (and stay in bounds of a familiar part). Backpropagated existing review items.
Rewritten the "decode" part to not rely on packed structs, and (subjectively) simplified the readout process.
Tested live on a MP15 board (DK1 STLink/V2-1 was not cooperating) to match both hex printout and decoded human-readable elements.

ALTracer commented:
Everything after first 4 commits is me moving the implementation into a common file and hooking (both existing and new) target commands to call it with a relevant UID base address (usually around OTP and F_SIZE). It might increase BMF size, but I believe it is done in an efficient enough manner.
Still, the "decoder" part can be disabled/suspended for stm32f1.c driver because the few chips I tested it on don't decode sensibly -- and the hex printout is still useful (for when you don't want to spin a gdb session just to grab 96 bits from a part-specific location outside of flash).
Passing by stm32l4 driver I replaced a couple `gdb_out()/gdb_outf()` with `tc_printf()` and removed `gdb_packet.h` which don't belong there (were used in mass-erase commands, and the proper way is O-packets).
TODO: needs rebasing on `main` and registering new TU in `meson.build`; optionally connecting with `utoa_upper()` of serialno.c, which is currently both static and platform-specific.

dragonmux commented on src/target/stm32_common.c:
```diff
@@ -0,0 +1,62 @@
+
+	for (size_t i = 0; i < 12U; i += 4U) {
+#if 0
+		const uint32_t value = read_le4(uid_bytes, i);
+		utoa_upper(value, &uid_hex[i], 16);
```
This looks so much nicer than the `snprintf()` monstrosity, please include making `utoa_upper()` a part of the hex utils in this PR and make use of it here!

ALTracer replied:
The migration will also require touching bootloader-related build system parts, so I'd rather drop the lines from here and do a separate PR.

dragonmux replied:
Given what this is, we encourage you to add hex_utils.c to the bootloader builds (CI will take care of verifying you didn't break anything) and use utoa_upper()..

dragonmux commented on src/target/stm32_common.c:
```diff
@@ -0,0 +1,62 @@
+	tc_printf(target, "0x%s\n", uid_hex);
+
+	stm32_uid_s uid;
+	//	memcpy(&uid, values, 12U);
```
Don't forget to remove this having changed to using buffer utils and the per-member value extraction.

ALTracer replied:
Removed. Not zero-initializing the stack-allocated struct, because I only ever use its fields, and not the undefined padding between them.

dragonmux replied:
Yep, that's all fine :D that actually would let you do this if you wanted:
```c
	stm32_uid_s uid = {
		.wafer_xcoord = read_le2(uid_bytes, 0),
		.wafer_ycoord = read_le2(uid_bytes, 2),
		.wafer_number = uid_bytes[4],
	};
	memcpy(uid.lot_number, &uid_bytes[5], 7);
```

dragonmux reviewed (commented):
We like the new approach much better, nice work! Only have the two review notes on this which are just re-iterating the conversation we already had on Discord about `utoa_upper()` really. We look forward to getting to merge this greatly!

ALTracer commented:
Rebased on main, hooked to meson.build, no other changes yet -- let's see whether it builds first

dragonmux reviewed (dismissed):
We've already noted about it being nice to switch the `snprintf()` out for `utoa_upper()` and simplify, but we understand your hesitation to do that change in this PR, so.. if you could please rebase this on main, we'll get this merged as it looks pretty sane otherwise and all review items have been addressed (with thanks!).

ALTracer commented:
> rebase this on main

Looks like there has been major changes to `target_mem*_read*()` API (PR1793) and medium changes to stm32h7.c, stm32f4.c (cosmetic `t`->`target` rename from PR1814) -- which give me the merge conflicts locally. While I understand the `s/target_mem_read32/target_mem32_read32/` change as present in merged PR1819, which function should I use for read_bytes for UID readout -- `s/target_mem_read/target_mem32_read/`?
Another problem is I replaced %02PRIX32 with %02PRIX8 in the commit before the move to stm32_common, because of -Wformat, and it works for BMDA; however for BMF it generates a %02hhX, but newlib-nano only (not newlib full, I tested both) omits support for the C99 char specifiers %hh and long long %ll, resulting in `(gdb) monitor uid` giving a constant string with no values: `0x0hX0hX0h0hX0hX0h0hX0hX0h`. Revert to PRIX32 and casts? `utoa` would obsolete this, but I'm not about to involve bootloader regress-testing (and meson learning) just yet.

dragonmux commented:
> While I understand the s/target_mem_read32/target_mem32_read32/ change as present in merged PR1819, which function should I use for read_bytes for UID readout -- s/target_mem_read/target_mem32_read/?

Yes, the insertion of the 32 is the address width used - these are 32-bit parts, so they're `mem32`. If they were 64-bit, they'd be `mem64`. The read width meaning of the suffix when present has not changed.

> Another problem is I replaced %02PRIX32 with %02PRIX8 in the commit before the move to stm32_common, because of -Wformat, and it works for BMDA; however for BMF it generates a %02hhX, but newlib-nano only (not newlib full, I tested both) omits support for the C99 char specifiers %hh and long long %ll, resulting in (gdb) monitor uid giving a constant string with no values: 0x0hX0hX0h0hX0hX0h0hX0hX0h. Revert to PRIX32 and casts?

Because those are formatting strings for uint8_t's, just `%02X` is fine, which solves for that problem in a reasonable enough way. It's not ideal, but it is how things are done elsewhere in the code base. As noted, once replaced with `utoa_upper()`, the problem with the warning permanently goes away anyway, so that's more than fine.

ALTracer commented:
> Because those are formatting strings for uint8_t's, just `%02X` is fine, which solves for that problem in a reasonable enough way. It's not ideal, but it is how things are done elsewhere in the code base.

I wrote this in one of commit messages:
> As snprintf is Wformat-checked (tc_printf was not), replace %02X with %02PRIX32.

However, I can't get it to trigger the warning? Maybe it was %08X for uint32_t `value & 0xffU`, then I refactored the code to do %02X for uint8_t `uid_bytes[i]`, now I'm at %08X for uint32_t `value` as-is -- it's already processed by `read_le4()` now that buffer utils exist.
If only `void tc_printf(target_s *target, const char *fmt, ...)` was marked `__attribute__((format(printf, 2, 3)))`...

Rebased to current main.

dragonmux commented:
> Maybe it was %08X for uint32_t value & 0xffU, then I refactored the code to do %02X for uint8_t uid_bytes[i]

This tracks with what we remember, and because it's now an array of `uint8_t`, you can just `%02X` them without worry.

dragonmux reviewed (approved):
LGTM, merging once the builds complete.
